src/server.rs

```diff
@@ -39,26 +42,43 @@ impl Server {
         metrics: Arc<Metrics>,
         rate_limiter: Arc<dyn RateLimit>,
         ip_addr_http_header: String,
+        api_keys: Vec<String>,
     ) -> Self {
         Self {
             listen_addr,
             registry,
             rate_limiter,
             metrics,
             ip_addr_http_header,
+            api_keys,
         }
     }
 
     pub async fn listen(&self, cancellation_token: CancellationToken) {
-        let router = Router::new()
-            .route("/healthz", get(healthz_handler))
-            .route("/ws", any(websocket_handler))
-            .with_state(ServerState {
-                registry: self.registry.clone(),
-                rate_limiter: self.rate_limiter.clone(),
-                metrics: self.metrics.clone(),
-                ip_addr_http_header: self.ip_addr_http_header.clone(),
-            });
+        let server_state = ServerState {
+            registry: self.registry.clone(),
+            rate_limiter: self.rate_limiter.clone(),
+            metrics: self.metrics.clone(),
+            ip_addr_http_header: self.ip_addr_http_header.clone(),
+            api_keys: self.api_keys.clone(),
+        };
+
+        let router = if self.api_keys.is_empty() {
+            info!(message = "API key authentication is disabled");
+            Router::new()
+                .route("/healthz", get(healthz_handler))
+                .route("/ws", any(websocket_handler))
+                .with_state(server_state)
```

Comment:
nit: You could move three of these four lines outside of conditional, and add the route for ws for the conditional.